lib/semantics/check-do.cc

```diff
@@ -64,6 +78,77 @@ class DoConcurrentBodyEnforce {
   template<typename T> bool Pre(const T &) { return true; }
   template<typename T> void Post(const T &) {}
 
+  template<typename T> bool Pre(const parser::Statement<T> &statement) {
+    currentStatementSourcePosition_ = statement.source;
+    if (statement.label.has_value()) {
+      labels_.insert(*statement.label);
+    }
+    return true;
+  }
+
+  // C1140 -- Can't deallocate a polymorphic entity in a DO CONCURRENT
+  // Note also that the deallocation of a derived type entity might cause the
+  // invocation of an IMPURE final subroutine.
+  void Post(const parser::BlockConstruct &blockConstruct) {
+    const auto &endBlockStmt{
+        std::get<parser::Statement<parser::EndBlockStmt>>(blockConstruct.t)};
+    const Scope &blockScope{context_.FindScope(endBlockStmt.source)};
+    const Scope &doScope{context_.FindScope(doConcurrentSourcePosition_)};
+    if (DoesScopeContain(&doScope, blockScope)) {
+      for (auto &pair : blockScope) {
+        Symbol &symbol{*pair.second};
+        if (IsPolymorphic(symbol) && IsAllocatable(symbol) &&
+            !symbol.attrs().test(Attr::SAVE)) {
+          SayWithDeclaration(context_, endBlockStmt.source,
+              "Deallocation of a polymorphic entity caused by block"
+              " exit not allowed in DO CONCURRENT"_err_en_US,
+              parser::CharBlock{}, symbol);
```

Comment:
Should this be a runtime check? How do we know that the allocatable variable was allocated?

Reply:
We don't know.  But note that constraint C1140 states that "A statement that might result in the deallocation of a polymorphic entity shall not appear within a DO CONCURRENT construct."  I interpret the use of the word "might" to say that we should check for this circumstance at compile time.

If this causes too many problems with existing Fortran code, we might want to downgrade this check to a warning.

Reply:
There isn't much extant Fortran that uses `DO CONCURRENT`.